File: back-end/src/domain/logic/pdf/script-direction.ts
export class ScriptDirection {
  /** Strong RTL characters: Arabic and Hebrew, base and presentation blocks. */
  private static readonly RTL_CHARS = /[\u0590-\u06FF\u0750-\u077F\u08A0-\u08FF\uFB1D-\uFDFF\uFE70-\uFEFF]/g;
  /** Strong LTR letters: Latin, Greek, Cyrillic, and CJK (all read left→right). */
  private static readonly LTR_CHARS = /[A-Za-z\u00C0-\u024F\u0370-\u03FF\u0400-\u04FF\u3040-\u30FF\u3400-\u9FFF]/g;

  isRightToLeft(text: string): boolean {
    const rtl = text.match(ScriptDirection.RTL_CHARS)?.length ?? 0;
    const ltr = text.match(ScriptDirection.LTR_CHARS)?.length ?? 0;
    return rtl > ltr;
  }
}

File: ocr-service/script_direction.py
"""Port of the backend ScriptDirection (domain/logic/pdf/script-direction.ts)."""

from __future__ import annotations

import re

# Strong RTL characters: Arabic and Hebrew, base and presentation blocks.
_RTL = re.compile(r"[֐-ۿݐ-ݿࢠ-ࣿיִ-﷿ﹰ-﻿]")
# Strong LTR letters: Latin, Greek, Cyrillic, and CJK.
_LTR = re.compile(r"[A-Za-zÀ-ɏͰ-ϿЀ-ӿ぀-ヿ㐀-鿿]")


def is_right_to_left(text: str) -> bool:
    return len(_RTL.findall(text)) > len(_LTR.findall(text))
